File: nfc_proof_of_concept/packages/nfc_host_card_emulation/android/src/main/kotlin/io/flutter/plugins/nfc_host_card_emulation/app_layer/ndef_format/serializers/ndef_message_serializer.kt
package io.flutter.plugins.nfc_host_card_emulation.ndef_format

import io.flutter.plugins.nfc_host_card_emulation.app_layer.ApduSerializer
import io.flutter.plugins.nfc_host_card_emulation.app_layer.Bytes
import io.flutter.plugins.nfc_host_card_emulation.ndef_format.fields.*
import io.flutter.plugins.nfc_host_card_emulation.ndef_format.serializers.NdefRecordSerializer
import java.io.ByteArrayOutputStream

class NdefFormatException(message: String) : Exception(message)

private enum class ChunkedRecordState {
    NOT_IN_CHUNK,
    AWAITING_MIDDLE_OR_END_CHUNK
}

private class NdefParser(private val rawMessage: Bytes) {
    private var offset = 0
    private var chunkState = ChunkedRecordState.NOT_IN_CHUNK
    private var reassembledPayload: ByteArrayOutputStream? = null
    private var reassembledType: NdefTypeField? = null
    private var reassembledId: NdefIdField? = null
    private var isFirstRecordInMessage = true

    fun parse(): List<NdefRecordSerializer> {
        val completeRecords = mutableListOf<NdefRecordSerializer>()
        while (offset < rawMessage.size) {
            val record = parseNextRecord()
            record?.let { completeRecords.add(it) }
            if (record?.flags?.isMessageEnd == true) break
        }

        if (chunkState == ChunkedRecordState.AWAITING_MIDDLE_OR_END_CHUNK) {
            throw NdefFormatException("Invalid NDEF message: message ends with an incomplete chunk sequence.")
        }

        return completeRecords
    }

    private fun parseNextRecord(): NdefRecordSerializer? {
        val flags = readFlags()
        val typeLength = readTypeLength(flags)
        val (payloadLength, payloadLengthBytes) = readPayloadLength(flags)
        val idLength = readIdLength(flags)
        val type = readType(flags, typeLength)
        val id = readId(flags, idLength)
        val payload = readPayload(payloadLength)

        return processRecord(flags, type, payload, id)
    }

    private fun processRecord(flags: NdefFlagByte, type: NdefTypeField, payload: NdefPayload?, id: NdefIdField?): NdefRecordSerializer? {
        when (chunkState) {
            ChunkedRecordState.NOT_IN_CHUNK -> {
                if (flags.isChunked) {
                    validateFirstChunk(flags)
                    chunkState = ChunkedRecordState.AWAITING_MIDDLE_OR_END_CHUNK
                    reassembledPayload = ByteArrayOutputStream().apply { payload?.buffer?.let { write(it) } }
                    reassembledType = type
                    reassembledId = id
                    return null
                } else {
                    val isLast = offset >= rawMessage.size
                    val record = NdefRecordSerializer.record(type, payload, id, isFirstRecordInMessage, isLast)
                    isFirstRecordInMessage = false
                    return record
                }
            }
            ChunkedRecordState.AWAITING_MIDDLE_OR_END_CHUNK -> {
                validateSubsequentChunk(flags)
                reassembledPayload?.write(payload?.buffer ?: byteArrayOf())
                if (flags.isChunked) {
                    return null
                } else {
                    chunkState = ChunkedRecordState.NOT_IN_CHUNK
                    val finalPayload = NdefPayload(reassembledPayload?.toByteArray() ?: byteArrayOf())
                    val record = NdefRecordSerializer.record(reassembledType!!, finalPayload, reassembledId, isFirstRecordInMessage, flags.isMessageEnd)
                    isFirstRecordInMessage = false
                    return record
                }
            }
        }
    }

    private fun readFlags(): NdefFlagByte {
        if (offset >= rawMessage.size) throw NdefFormatException("Unexpected end of data when reading flags.")
        return NdefFlagByte.fromByte(rawMessage[offset++])
    }

    private fun readTypeLength(flags: NdefFlagByte): Int {
        if (flags.tnf == Tnf.EMPTY || (chunkState == ChunkedRecordState.AWAITING_MIDDLE_OR_END_CHUNK && flags.tnf == Tnf.UNCHANGED)) return 0
        if (offset >= rawMessage.size) throw NdefFormatException("Unexpected end of data when reading type length.")
        return rawMessage[offset++].toInt() and 0xFF
    }

    private fun readPayloadLength(flags: NdefFlagByte): Pair<Long, Int> {
        val (field, bytesRead) = NdefPayloadLengthField.fromBytes(flags.isShortRecord, rawMessage, offset)
        offset += bytesRead
        return Pair(field.getValue(), bytesRead)
    }

    private fun readIdLength(flags: NdefFlagByte): Int {
        if (!flags.hasId) return 0
        if (chunkState == ChunkedRecordState.AWAITING_MIDDLE_OR_END_CHUNK) throw NdefFormatException("Invalid chunk: ID length must only be present in the first chunk.")
        if (offset >= rawMessage.size) throw NdefFormatException("Unexpected end of data when reading ID length.")
        return rawMessage[offset++].toInt() and 0xFF
    }

    private fun readType(flags: NdefFlagByte, length: Int): NdefTypeField {
        if (length == 0) return NdefTypeField.fromBytes(byteArrayOf(), flags.tnf)
        if (chunkState == ChunkedRecordState.AWAITING_MIDDLE_OR_END_CHUNK) throw NdefFormatException("Invalid chunk: Type must only be present in the first chunk.")
        if (offset + length > rawMessage.size) throw NdefFormatException("Unexpected end of data when reading type.")
        val typeBytes = rawMessage.sliceArray(offset until offset + length)
        offset += length
        return NdefTypeField.fromBytes(typeBytes, flags.tnf)
    }
    
    private fun readId(flags: NdefFlagByte, length: Int): NdefIdField? {
        if (!flags.hasId || length == 0) return null
        if (offset + length > rawMessage.size) throw NdefFormatException("Unexpected end of data when reading ID.")
        val idBytes = rawMessage.sliceArray(offset until offset + length)
        offset += length
        return NdefIdField(idBytes)
    }

    private fun readPayload(length: Long): NdefPayload? {
        if (length == 0L) return null
        if (offset + length > rawMessage.size) throw NdefFormatException("Unexpected end of data when reading payload.")
        val payloadBytes = rawMessage.sliceArray(offset until offset + length.toInt())
        offset += length.toInt()
        return NdefPayload(payloadBytes)
    }

    private fun validateFirstChunk(flags: NdefFlagByte) {
        if (flags.tnf == Tnf.EMPTY || flags.tnf == Tnf.UNCHANGED) {
            throw NdefFormatException("Invalid first chunk: TNF cannot be Empty or Unchanged.")
        }
    }

    private fun validateSubsequentChunk(flags: NdefFlagByte) {
        if (flags.tnf != Tnf.UNCHANGED) {
            throw NdefFormatException("Invalid subsequent chunk: TNF must be Unchanged.")
        }
        if (flags.hasId) {
            throw NdefFormatException("Invalid subsequent chunk: ID field is not allowed.")
        }
    }
}

data class NdefRecordData(val type: NdefTypeField, val payload: NdefPayload?, val id: NdefIdField?)

class NdefMessageSerializer private constructor(private val records: List<NdefRecordSerializer>) : ApduSerializer("NDEF Message") {
    fun getRecordsData(): List<NdefRecordData> = records.map { it.toData() }

    companion object {
        @JvmStatic
        fun fromBytes(rawMessage: Bytes): NdefMessageSerializer {
            require(rawMessage.isNotEmpty()) { "Cannot parse an empty NDEF message." }
            val parser = NdefParser(rawMessage)
            val parsedRecords = parser.parse()
            require(parsedRecords.isNotEmpty()) { "Failed to parse any valid records from the NDEF message." }
            return NdefMessageSerializer(parsedRecords)
        }

        @JvmStatic
        fun fromRecords(recordData: List<NdefRecordData>): NdefMessageSerializer {
            require(recordData.isNotEmpty()) { "Cannot create an NDEF message with zero records." }
            val serializedRecords = recordData.mapIndexed { index, data ->
                NdefRecordSerializer.record(
                    type = data.type,
                    payload = data.payload,
                    id = data.id,
                    isFirstInMessage = (index == 0),
                    isLastInMessage = (index == recordData.lastIndex)
                )
            }
            return NdefMessageSerializer(serializedRecords)
        }
    }

    override fun setFields() {
        fields = records
    }
}

File: nfc_proof_of_concept/packages/flutter_hce/lib/app_layer/ndef_format/serializers/ndef_message_serializer.dart
import 'dart:typed_data';

import '../../field.dart';
import '../fields/ndef_record_fields.dart';
import 'ndef_record_serializer.dart';

class NdefFormatException implements Exception {
  final String message;
  NdefFormatException(this.message);
  @override
  String toString() => 'NdefFormatException: $message';
}

enum _ChunkedRecordState { notInChunk, awaitingMiddleOrEndChunk }


class _NdefParser {
  final Uint8List _rawMessage;
  int _offset = 0;
  _ChunkedRecordState _chunkState = _ChunkedRecordState.notInChunk;
  BytesBuilder? _reassembledPayload;
  NdefTypeField? _reassembledType;
  NdefIdField? _reassembledId;
  bool _isFirstRecordInMessage = true;

  _NdefParser(this._rawMessage);

  List<NdefRecordSerializer> parse() {
    final completeRecords = <NdefRecordSerializer>[];
    while (_offset < _rawMessage.length) {
      final record = _parseNextRecord();
      if (record != null) {
        completeRecords.add(record);
      }
      if (record?.flags.isMessageEnd ?? false) break;
    }

    if (_chunkState == _ChunkedRecordState.awaitingMiddleOrEndChunk) {
      throw NdefFormatException('Invalid NDEF message: message ends with an incomplete chunk sequence.');
    }
    return completeRecords;
  }

  NdefRecordSerializer? _parseNextRecord() {
    final flags = _readFlags();
    final typeLength = _readTypeLength(flags);
    final (payloadLength, payloadLengthBytes) = _readPayloadLength(flags);
    final idLength = _readIdLength(flags);
    final type = _readType(flags, typeLength);
    final id = _readId(flags, idLength);
    final payload = _readPayload(payloadLength);

    return _processRecord(flags, type, payload, id);
  }

  NdefRecordSerializer? _processRecord(NdefFlagByte flags, NdefTypeField type, NdefPayload? payload, NdefIdField? id) {
    switch (_chunkState) {
      case _ChunkedRecordState.notInChunk:
        if (flags.isChunked) {
          _validateFirstChunk(flags);
          _chunkState = _ChunkedRecordState.awaitingMiddleOrEndChunk;
          _reassembledPayload = BytesBuilder(copy: false)..add(payload?.buffer ?? Uint8List(0));
          _reassembledType = type;
          _reassembledId = id;
          return null; 
        } else {
          final isLast = _offset >= _rawMessage.length;
          final record = NdefRecordSerializer.record(type: type, payload: payload, id: id, isFirstInMessage: _isFirstRecordInMessage, isLastInMessage: isLast);
          _isFirstRecordInMessage = false;
          return record;
        }
      case _ChunkedRecordState.awaitingMiddleOrEndChunk:
        _validateSubsequentChunk(flags);
        _reassembledPayload?.add(payload?.buffer ?? Uint8List(0));
        if (flags.isChunked) {
          return null;
        } else {
          _chunkState = _ChunkedRecordState.notInChunk;
          final finalPayload = NdefPayload(_reassembledPayload?.takeBytes() ?? Uint8List(0));
          final record = NdefRecordSerializer.record(type: _reassembledType!, payload: finalPayload, id: _reassembledId, isFirstInMessage: _isFirstRecordInMessage, isLastInMessage: flags.isMessageEnd);
          _isFirstRecordInMessage = false;
          return record;
        }
    }
  }

  NdefFlagByte _readFlags() {
    if (_offset >= _rawMessage.length) throw NdefFormatException("Unexpected end of data when reading flags.");
    return NdefFlagByte.fromByte(_rawMessage[_offset++]);
  }

  int _readTypeLength(NdefFlagByte flags) {
    if (flags.tnf == Tnf.empty || (_chunkState == _ChunkedRecordState.awaitingMiddleOrEndChunk && flags.tnf == Tnf.unchanged)) return 0;
    if (_offset >= _rawMessage.length) throw NdefFormatException("Unexpected end of data when reading type length.");
    return _rawMessage[_offset++];
  }

  (int, int) _readPayloadLength(NdefFlagByte flags) {
    if (flags.isShortRecord) {
      if (_offset >= _rawMessage.length) throw NdefFormatException("Unexpected end of data for short payload length.");
      return (_rawMessage[_offset++], 1);
    } else {
      if (_offset + 3 >= _rawMessage.length) throw NdefFormatException("Unexpected end of data for long payload length.");
      final length = ByteData.view(_rawMessage.buffer).getUint32(_offset, Endian.big);
      _offset += 4;
      return (length, 4);
    }
  }

  int _readIdLength(NdefFlagByte flags) {
    if (!flags.hasId) return 0;
    if (_chunkState == _ChunkedRecordState.awaitingMiddleOrEndChunk) throw NdefFormatException("Invalid chunk: ID length must only be present in the first chunk.");
    if (_offset >= _rawMessage.length) throw NdefFormatException("Unexpected end of data when reading ID length.");
    return _rawMessage[_offset++];
  }

  NdefTypeField _readType(NdefFlagByte flags, int length) {
    if (length == 0) return NdefTypeField(Uint8List(0), tnf: flags.tnf);
    if (_chunkState == _ChunkedRecordState.awaitingMiddleOrEndChunk) throw NdefFormatException("Invalid chunk: Type must only be present in the first chunk.");
    if (_offset + length > _rawMessage.length) throw NdefFormatException("Unexpected end of data when reading type.");
    final typeBytes = _rawMessage.sublist(_offset, _offset + length);
    _offset += length;
    return NdefTypeField(typeBytes, tnf: flags.tnf);
  }

  NdefIdField? _readId(NdefFlagByte flags, int length) {
    if (!flags.hasId || length == 0) return null;
    if (_offset + length > _rawMessage.length) throw NdefFormatException("Unexpected end of data when reading ID.");
    final idBytes = _rawMessage.sublist(_offset, _offset + length);
    _offset += length;
    return NdefIdField(idBytes);
  }
  
  NdefPayload? _readPayload(int length) {
    if (length == 0) return null;
    if (_offset + length > _rawMessage.length) throw NdefFormatException("Unexpected end of data when reading payload.");
    final payloadBytes = _rawMessage.sublist(_offset, _offset + length);
    _offset += length;
    return NdefPayload(payloadBytes);
  }

  void _validateFirstChunk(NdefFlagByte flags) {
    if (flags.tnf == Tnf.empty || flags.tnf == Tnf.unchanged) {
      throw NdefFormatException("Invalid first chunk: TNF cannot be Empty or Unchanged.");
    }
  }

  void _validateSubsequentChunk(NdefFlagByte flags) {
    if (flags.tnf != Tnf.unchanged) {
      throw NdefFormatException("Invalid subsequent chunk: TNF must be Unchanged.");
    }
    if (flags.hasId) {
      throw NdefFormatException("Invalid subsequent chunk: ID field is not allowed.");
    }
  }
}

class NdefMessageSerializer extends ApduSerializer {
  final List<NdefRecordSerializer> records;

  factory NdefMessageSerializer.fromBytes(Uint8List rawMessage) {
    if (rawMessage.isEmpty) {
      throw ArgumentError('Cannot parse an empty NDEF message.');
    }
    final parser = _NdefParser(rawMessage);
    final records = parser.parse();
    if (records.isEmpty) {
      throw ArgumentError('Failed to parse any valid records from the NDEF message.');
    }
    return NdefMessageSerializer._internal(records);
  }

  factory NdefMessageSerializer.fromRecords({
    required List<NdefRecordSerializer> records,
  }) {
    if (records.isEmpty) {
      throw ArgumentError('Cannot create an NDEF message with zero records.');
    }

    return NdefMessageSerializer._internal(records);
  }

  NdefMessageSerializer._internal(this.records) : super(name: "NDEF Message");

  @override
  void setFields() {
    fields = records;
  }
}